soco/music_services/music_service.py

```diff
@@ -155,50 +196,138 @@ def call(self, method, args=None):
                     http_headers=self.http_headers,
-                    soap_action="http://www.sonos.com/Services/1"
-                    ".1#{}".format(method),
+                    soap_action="http://www.sonos.com/Services/1.1#{0}".format(method),
                     soap_header=self.get_soap_header(),
                     namespace=self.namespace,
                     timeout=self.timeout,
                 )
                 result_elt = message.call()
-
             else:
+                log.exception(
+                    "Unhandled SOAP Fault. Code: %s. Detail: %s. String: %s",
+                    exc.faultcode,
+                    exc.detail,
+                    exc.faultstring,
+                )
                 raise MusicServiceException(exc.faultstring, exc.faultcode) from exc
+        except XML.ParseError as parse_exc:
+            raise MusicServiceAuthException(
+                "Got empty response to request, likely because the service is not "
+                "authenticated"
+            ) from parse_exc
 
         # The top key in the OrderedDict will be the methodResult. Its
         # value may be None if no results were returned.
         result = list(
             parse(
                 XML.tostring(result_elt),
                 process_namespaces=True,
-                namespaces={"http://www.sonos.com/Services/1.1": None},
+                namespaces={self.namespace: None},
             ).values()
         )[0]
 
         return result if result is not None else {}
 
+    def device_or_app_link_auth_part1(self):
+        """Perform part 1 of a Device or App Link authentication session
+
+        See `MusicService.device_or_app_link_auth_part1` for details
+
+        """
+        link_device_id = None
+
+        if self.music_service.auth_type == "DeviceLink":
+            log.debug("Perform DeviceLink auth part 1")
+            result = self.call(
+                "getDeviceLinkCode", [("householdId", self._household_id)]
+            )["getDeviceLinkCodeResult"]
+            if "linkDeviceId" in result:
+                link_device_id = result["linkDeviceId"]
+            return result["regUrl"], result["linkCode"], link_device_id
+        elif self.music_service.auth_type == "AppLink":
+            log.debug("Perform AppLink auth part 1")
+            result = self.call("getAppLink", [("householdId", self._household_id)])[
+                "getAppLinkResult"
+            ]
+            auth_parts = result["authorizeAccount"]["deviceLink"]
+            if "linkDeviceId" in auth_parts:
+                link_device_id = auth_parts["linkDeviceId"]
+            return auth_parts["regUrl"], auth_parts["linkCode"], link_device_id
+        raise MusicServiceAuthException(
+            "device_or_app_link_auth_part1 is not implemented "
+            "for auth type {}".format(self.music_service.auth_type)
+        )
+
+    def device_or_app_link_auth_part2(self, link_code, link_device_id=None):
```

Comment:
This (and its equivalent in `MusicService`) should probably return a `bool` for success.

Reply:
I don't think so. For Python I don't think it is very common or idiomatic to return success or failure. If it doesn't raise an exception it succeeded.